src/pybamm/solvers/base_processed_variable.py

```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
+
+import abc
+
+import pybamm
+
+
+class BaseProcessedVariable(abc.ABC):
+    """
+    Shared API for both 'on-demand' and 'computed' processed variables.
+    """
+
+    @abc.abstractmethod
+    def __call__(self, **coords): ...
+
+    @abc.abstractmethod
+    def as_computed(self) -> pybamm.ProcessedVariableComputed: ...
+
+    def _update(
```

Comment:
is this meant to be a private or public method? you use it outside the class so we might want to remove the underscore to make this clear
```suggestion
    def update(
```